File: server/src/main/kotlin/com/example/reconix/server/service/EmailService.kt
package com.example.reconix.server.service

import jakarta.mail.*
import jakarta.mail.internet.InternetAddress
import jakarta.mail.internet.MimeBodyPart
import jakarta.mail.internet.MimeMessage
import jakarta.mail.internet.MimeMultipart
import jakarta.mail.search.FlagTerm
import kotlinx.coroutines.*
import java.io.File
import java.util.Properties

/**
 * Email Service - Handles SMTP sending and IMAP email monitoring
 *
 * Environment Variables:
 * - SMTP_HOST, SMTP_PORT, SMTP_USER, SMTP_PASSWORD
 * - IMAP_HOST, IMAP_PORT, IMAP_USER, IMAP_PASSWORD
 *
 * All methods gracefully no-op if credentials are not configured.
 */
class EmailService {

    private val smtpHost = System.getenv("SMTP_HOST")
    private val smtpPort = System.getenv("SMTP_PORT") ?: "587"
    private val smtpUser = System.getenv("SMTP_USER")
    private val smtpPassword = System.getenv("SMTP_PASSWORD")

    private val imapHost = System.getenv("IMAP_HOST")
    private val imapPort = System.getenv("IMAP_PORT") ?: "993"
    private val imapUser = System.getenv("IMAP_USER")
    private val imapPassword = System.getenv("IMAP_PASSWORD")

    private val uploadDir = "uploads/invoices"
    private var emailListenerJob: Job? = null

    val isSmtpConfigured: Boolean get() = !smtpHost.isNullOrBlank() && !smtpUser.isNullOrBlank()
    val isImapConfigured: Boolean get() = !imapHost.isNullOrBlank() && !imapUser.isNullOrBlank()

    init {
        File(uploadDir).mkdirs()
    }

    /**
     * Send PO PDF to vendor via email
     * @return true if email sent successfully, false if not configured or failed
     */
    fun sendPoToVendor(vendorEmail: String, poId: String, pdfPath: String): Boolean {
        if (!isSmtpConfigured) {
            println("📧 [SMTP NOT CONFIGURED] Would send PO $poId to $vendorEmail")
            println("   Set SMTP_HOST, SMTP_USER, SMTP_PASSWORD environment variables to enable")
            return false
        }

        return try {
            val props = Properties().apply {
                put("mail.smtp.host", smtpHost!!)
                put("mail.smtp.port", smtpPort)
                put("mail.smtp.auth", "true")
                put("mail.smtp.starttls.enable", "true")
            }

            val session = Session.getInstance(props, object : Authenticator() {
                override fun getPasswordAuthentication(): PasswordAuthentication {
                    return PasswordAuthentication(smtpUser, smtpPassword)
                }
            })

            val message = MimeMessage(session).apply {
                setFrom(InternetAddress(smtpUser, "Reconix Procurement"))
                setRecipient(Message.RecipientType.TO, InternetAddress(vendorEmail))
                subject = "Purchase Order $poId - Reconix"
            }

            // Build multipart message: text body + PDF attachment
            val multipart = MimeMultipart()

            // Text body
            val textPart = MimeBodyPart()
            textPart.setContent("""
                <html>
                <body>
                <h2>Purchase Order $poId</h2>
                <p>Dear Vendor,</p>
                <p>Please find attached the Purchase Order <strong>$poId</strong> for your reference.</p>
                <p>Kindly review the order details and send your invoice referencing this PO number.</p>
                <br>
                <p>Best regards,<br>Reconix Procurement Team</p>
                </body>
                </html>
            """.trimIndent(), "text/html")
            multipart.addBodyPart(textPart)

            // PDF attachment
            val pdfFile = File(pdfPath)
            if (pdfFile.exists()) {
                val attachmentPart = MimeBodyPart()
                attachmentPart.attachFile(pdfFile)
                attachmentPart.fileName = "$poId.pdf"
                multipart.addBodyPart(attachmentPart)
            }

            message.setContent(multipart)
            Transport.send(message)

            println("📧 ✅ Email sent: PO $poId -> $vendorEmail")
            true
        } catch (e: Exception) {
            println("📧 ❌ Failed to send email: ${e.message}")
            false
        }
    }

    /**
     * Start the email listener background service
     * Checks for new invoice emails every 5 minutes
     * @param onInvoiceReceived Callback when an invoice PDF is found
     */
    fun startEmailListener(
        scope: CoroutineScope,
        onInvoiceReceived: (fileName: String, fileBytes: ByteArray) -> Unit
    ) {
        if (!isImapConfigured) {
            println("📬 [IMAP NOT CONFIGURED] Email listener not started")
            println("   Set IMAP_HOST, IMAP_USER, IMAP_PASSWORD environment variables to enable")
            return
        }

        emailListenerJob = scope.launch(Dispatchers.IO) {
            println("📬 Email listener started - checking every 5 minutes")
            while (isActive) {
                try {
                    checkForNewEmails(onInvoiceReceived)
                } catch (e: Exception) {
                    println("📬 ❌ Email check error: ${e.message}")
                }
                delay(5 * 60 * 1000L) // 5 minutes
            }
        }
    }

    /**
     * Check IMAP mailbox for new emails with PDF attachments
     */
    private fun checkForNewEmails(onInvoiceReceived: (String, ByteArray) -> Unit) {
        val props = Properties().apply {
            put("mail.imap.host", imapHost!!)
            put("mail.imap.port", imapPort)
            put("mail.imap.ssl.enable", "true")
        }

        val session = Session.getInstance(props)
        val store = session.getStore("imaps")

        try {
            store.connect(imapHost, imapUser, imapPassword)
            val inbox = store.getFolder("INBOX")
            inbox.open(Folder.READ_WRITE)

            // Search for unread messages
            val unreadMessages = inbox.search(FlagTerm(Flags(Flags.Flag.SEEN), false))
            println("📬 Found ${unreadMessages.size} unread emails")

            for (message in unreadMessages) {
                processEmailMessage(message, onInvoiceReceived)
                // Mark as read
                message.setFlag(Flags.Flag.SEEN, true)
            }

            inbox.close(false)
        } finally {
            store.close()
        }
    }

    /**
     * Process a single email message - extract PDF attachments
     */
    private fun processEmailMessage(
        message: Message,
        onInvoiceReceived: (String, ByteArray) -> Unit
    ) {
        val content = message.content

        if (content is Multipart) {
            for (i in 0 until content.count) {
                val bodyPart = content.getBodyPart(i)
                val disposition = bodyPart.disposition

                if (disposition != null &&
                    (disposition.equals(Part.ATTACHMENT, ignoreCase = true) ||
                     disposition.equals(Part.INLINE, ignoreCase = true))
                ) {
                    val fileName = bodyPart.fileName ?: "unknown_${System.currentTimeMillis()}.pdf"
                    if (fileName.lowercase().endsWith(".pdf") ||
                        fileName.lowercase().endsWith(".png") ||
                        fileName.lowercase().endsWith(".jpg") ||
                        fileName.lowercase().endsWith(".jpeg")
                    ) {
                        val fileBytes = bodyPart.inputStream.readBytes()
                        println("📬 📎 Found attachment: $fileName (${fileBytes.size} bytes)")

                        // Save to disk
                        val savedPath = "$uploadDir/$fileName"
                        File(savedPath).writeBytes(fileBytes)

                        // Trigger processing callback
                        onInvoiceReceived(fileName, fileBytes)
                    }
                }
            }
        }
    }

    fun stopEmailListener() {
        emailListenerJob?.cancel()
        emailListenerJob = null
        println("📬 Email listener stopped")
    }
}

File: python_server/services/email_service.py
"""
services/email_service.py - Send PO emails via SMTP, receive invoices via IMAP
Equivalent to server/.../service/EmailService.kt
"""
import imaplib
import os
import smtplib
import threading
import time
from email import message_from_bytes
from email.mime.application import MIMEApplication
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from typing import Callable, Optional

SMTP_HOST = os.getenv("SMTP_HOST", "")
SMTP_PORT = int(os.getenv("SMTP_PORT", "587"))
SMTP_USER = os.getenv("SMTP_USER", "")
SMTP_PASSWORD = os.getenv("SMTP_PASSWORD", "")

IMAP_HOST = os.getenv("IMAP_HOST", "")
IMAP_PORT = int(os.getenv("IMAP_PORT", "993"))
IMAP_USER = os.getenv("IMAP_USER", "")
IMAP_PASSWORD = os.getenv("IMAP_PASSWORD", "")

UPLOAD_DIR = "uploads/invoices"
os.makedirs(UPLOAD_DIR, exist_ok=True)

_is_smtp_configured = bool(SMTP_HOST and SMTP_USER)
_is_imap_configured = bool(IMAP_HOST and IMAP_USER)

_listener_thread: Optional[threading.Thread] = None
_listener_stop_event = threading.Event()


def send_po_to_vendor(vendor_email: str, po_id: str, pdf_path: str) -> bool:
    if not _is_smtp_configured:
        print(f"📧 [SMTP NOT CONFIGURED] Would send PO {po_id} to {vendor_email}")
        return False
    try:
        msg = MIMEMultipart()
        msg["From"] = f"Reconix Procurement <{SMTP_USER}>"
        msg["To"] = vendor_email
        msg["Subject"] = f"Purchase Order {po_id} - Reconix"

        html_body = f"""
        <html><body>
        <h2>Purchase Order {po_id}</h2>
        <p>Dear Vendor,</p>
        <p>Please find attached the Purchase Order <strong>{po_id}</strong> for your reference.</p>
        <p>Kindly review the order details and send your invoice referencing this PO number.</p>
        <br><p>Best regards,<br>Reconix Procurement Team</p>
        </body></html>
        """
        msg.attach(MIMEText(html_body, "html"))

        if os.path.exists(pdf_path):
            with open(pdf_path, "rb") as f:
                attachment = MIMEApplication(f.read(), _subtype="pdf")
                attachment.add_header("Content-Disposition", "attachment", filename=f"{po_id}.pdf")
                msg.attach(attachment)

        with smtplib.SMTP(SMTP_HOST, SMTP_PORT) as server:
            server.ehlo()
            server.starttls()
            server.login(SMTP_USER, SMTP_PASSWORD)
            server.sendmail(SMTP_USER, vendor_email, msg.as_string())

        print(f"📧 ✅ Email sent: PO {po_id} -> {vendor_email}")
        return True
    except Exception as exc:
        print(f"📧 ❌ Failed to send email: {exc}")
        return False


def start_email_listener(on_invoice_received: Callable[[str, bytes], None]) -> None:
    if not _is_imap_configured:
        print("📬 [IMAP NOT CONFIGURED] Email listener not started")
        return
    _listener_stop_event.clear()
    thread = threading.Thread(
        target=_email_listener_loop,
        args=(on_invoice_received,),
        daemon=True,
        name="EmailListener",
    )
    thread.start()
    global _listener_thread
    _listener_thread = thread
    print("📬 Email listener started - checking every 5 minutes")


def stop_email_listener() -> None:
    _listener_stop_event.set()
    global _listener_thread
    _listener_thread = None
    print("📬 Email listener stopped")


def _email_listener_loop(on_invoice_received: Callable[[str, bytes], None]) -> None:
    while not _listener_stop_event.is_set():
        try:
            _check_for_new_emails(on_invoice_received)
        except Exception as exc:
            print(f"📬 ❌ Email check error: {exc}")
        _listener_stop_event.wait(timeout=5 * 60)


def _check_for_new_emails(on_invoice_received: Callable[[str, bytes], None]) -> None:
    mail = imaplib.IMAP4_SSL(IMAP_HOST, IMAP_PORT)
    try:
        mail.login(IMAP_USER, IMAP_PASSWORD)
        mail.select("INBOX")
        _, data = mail.search(None, "UNSEEN")
        ids = data[0].split()
        print(f"📬 Found {len(ids)} unread emails")
        for uid in ids:
            _, msg_data = mail.fetch(uid, "(RFC822)")
            raw = msg_data[0][1]
            _process_email_message(raw, on_invoice_received)
            mail.store(uid, "+FLAGS", "\\Seen")
    finally:
        mail.logout()


def _process_email_message(raw: bytes, on_invoice_received: Callable[[str, bytes], None]) -> None:
    msg = message_from_bytes(raw)
    if msg.is_multipart():
        for part in msg.walk():
            content_disposition = part.get("Content-Disposition", "")
            if content_disposition.lower().startswith(("attachment", "inline")):
                file_name = part.get_filename() or f"unknown_{int(time.time())}.pdf"
                lower_name = file_name.lower()
                if any(lower_name.endswith(ext) for ext in (".pdf", ".png", ".jpg", ".jpeg")):
                    file_bytes = part.get_payload(decode=True) or b""
                    print(f"📬 📎 Found attachment: {file_name} ({len(file_bytes)} bytes)")
                    save_path = os.path.join(UPLOAD_DIR, file_name)
                    with open(save_path, "wb") as f:
                        f.write(file_bytes)
                    on_invoice_received(file_name, file_bytes)
